drivers/at24cxxx/mtd/mtd.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2020 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ *
+ */
+
+/**
+ * @ingroup     drivers_at24cxxx
+ * @{
+ *
+ * @file
+ * @brief       MTD EEPROM driver implementation for at24cxxx EEPROM
+ *
+ * @author      Fabian Hüßler <fabian.huessler@ovgu.de>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <assert.h>
+
+#include "mtd_at24cxxx.h"
+
+#define DEV(mtd_ptr)        (((mtd_at24cxxx_t *)(mtd_ptr))->at24cxxx_eeprom)
+#define PARAMS(mtd_ptr)     (((mtd_at24cxxx_t *)(mtd_ptr))->params)
+
+static int _mtd_at24cxxx_init(mtd_dev_t *mtd)
+{
+    assert(mtd);
+    assert(mtd->driver == &mtd_at24cxxx_driver);
+    assert(DEV(mtd));
+    assert(PARAMS(mtd));
+    int init = at24cxxx_init(DEV(mtd), PARAMS(mtd));
+    if (init != AT24CXXX_OK) {
+        return init;
+    }
+    mtd->page_size = DEV(mtd)->params.page_size;
+    mtd->pages_per_sector = 1;
+    mtd->sector_count = DEV(mtd)->params.eeprom_size /
+                        DEV(mtd)->params.page_size;
+    return 0;
+}
+
+static int _mtd_at24cxxx_read(mtd_dev_t *mtd, void *dest, uint32_t addr,
+                              uint32_t size)
+{
+    return at24cxxx_read(DEV(mtd), addr, dest, size) == AT24CXXX_OK ? 0 : -EIO;
```

Comment:
You could just make the `at24cxxx` functions return 0 or `-EIO` on their own.

Reply:
Are you sure this makes a difference?
```
at24cxxx_read(...) {
...
    return check == AT24CXXX_OK ? check : -EIO;
}
```
It just propagates the value checking to a lower level, while the requirement to return `-EIO` exists on `MTD` level.

Reply:
I was hoping clever LTO could throw out the wrapper functions if we just pass on the return values, but I did not try this.

Reply:
`at24cxxx_read()` will either return 0 or `-ERANGE`, so you could just 

    return at24cxxx_read(DEV(mtd), addr, dest, size);

here and comply with the API.

Reply:
`at24cxxx_read()` will pass through errors from `i2c_read_regs()`.
Of course I´d rather just like to do the function call.
But what if someone relies on `-EIO` explicitly? Altough, I did not find it in existing code.
If you say that´s not a problem for the benefit of  LTO, then I´ll change.

Reply:
tbh I was also hoping to get some consistency that way, e.g. that the user can expect the same error values from MTD and the device itself.

right now `mtd_read()` will return

> `-EOVERFLOW if addr or count are not valid, i.e. outside memory`

and `at24cxxx_read()` returns 

> `-ERANGE if pos + len is out of bounds`

Maybe there is a reason for that, but at a glance this looks inconsistent.